cde/lib/tt/man/man1/ttsnoop.1
.de LI
.\" simulate -mm .LIs by turning them into .TPs
.TP \\n()Jn
\\$1
..
.TH ttsnoop 1 "1 March 1996" "ToolTalk 1.3" "ToolTalk Commands" 
.BH "1 March 1996"
.\" (c) Copyright 1993, 1994 Hewlett-Packard Company
.\" (c) Copyright 1993, 1994 International Business Machines Corp.
.\" (c) Copyright 1993, 1994 Sun Microsystems, Inc.
.\" (c) Copyright 1993, 1994 Novell, Inc.
.IX "ttsnoop.1" "" "ttsnoop.1" ""
.SH NAME
ttsnoop \- ToolTalk graphical user interface
.SH SYNOPSIS
.ft 3
.fi
.na
ttsnoop
[\f2options\fP]
[\(miF\0\f2scopefile\fP]
[\(mi<\0\f2procid\fP]
[\(miv\0\f2media\fP]
[\(mim\0\f2op\fP]
.PP
.fi
.ft 3
.fi
.na
ttsnoop
[\f2options\fP]
[\(mie\0\f2script\fP]
command
[\f2args\fP]
.PP
.fi
.ft 3
.fi
.na
ttsnoop
[\f2options\fP]
\(min
\(bv
\(miN
.PP
.fi
.ft 3
.fi
.na
ttsnoop
[\(miTuX]
[\(miS\0\f2sessid\fP]
[\(miw\0\f2n\fP]
[\(mil\0\f2n\fP]
[\(mio\0\f2snoopfile\fP]
[\(miO\0\f2tracefile\fP]
.PP
.fi
.SH DESCRIPTION
.LP
The
.BR ttsnoop 
utility interactively monitors ToolTalk message traffic,
.BR ttsession (1)
pattern matching, and ToolTalk client function calls.
.BR ttsnoop 
allows the interactive execution of almost any valid sequence of
ToolTalk function calls, while optionally tracing those calls.
.BR ttsnoop 
can interactively create and send any ToolTalk message, and
can automatically create many of the standard ToolTalk messages.
.BR ttsnoop 
can interactively create and register any ToolTalk pattern.  Messages
received by virtue of these patterns can be processed (e.g. replied to)
interactively or automatically.  For any message encountered,
.BR ttsnoop 
can clone a copy of it or generate a pattern that will match similar messages.
For any message encountered,
.BR ttsnoop 
can generate C source code for recreating it,
.BR dtactionfile (4)
source for a message action that will send it, or
.BR tt_type_comp (1)
source for a static pattern that will match it.
.BR ttsnoop 
can dump the system's installed static patterns and message actions.
.PP
Like
.BR tttrace (1),
.BR ttsnoop 
operates in two fundamental modes.
.BR ttsnoop 
either runs
.BR command 
with ToolTalk client tracing turned on, or (if
.BR command 
is omitted) snoops message traffic in the default ToolTalk scope.
For client tracing,
.BR ttsnoop 
simply invokes
.BR tttrace 
and logs the trace output to the
.BR ttsnoop 
terminal pane.
For message snooping,
.BR ttsnoop 
registers a ToolTalk pattern in the default scope and prints each
matching message in the terminal pane.  When
.BR ttsnoop 
receives a message, it prints the current time, the address
of the internal
.BR Tt_message_callback 
that received the message, and a description of the message as
generated by
.BR tt_message_print (3).
.SH "OPTIONS"
The following options are available:
.PP
.RS 3
.nr )J 6
.LI \f3\(miF\0\f2scopefile\f1
.br
Scope initial pattern also to
.IR scopefile .
.LI \f3\(mi<\0\f2procid\f1
.br
Limit initial pattern to messages from
.IR procid .
.LI \f3\(miv\0\f2mediaType\f1
.br
Limit initial pattern to messages for
.IR mediaType .
.LI \f3\(mim\0\f2op\f1
.br
Limit initial pattern to messages with
.IR op .
.LI \f3\(min\fP
Skip initial pattern.
.LI \f3\(miN\fP
Skip initial
.BR ttdt_open (3),
also.
.LI \f3\(mie\0\f2script\f1
.br
Take
.IR script
as a
.BR tttrace (1)
setting.  See
.BR tttracefile (4).
.LI \f3\(miT\fP
Trace (even initial)
ToolTalk API calls made by
.BR ttsnoop. 
.LI \f3\(miu\fP
Map (de-iconify) on snoop output.
.LI \f3\(miS\0\f2sessid\f1
.br
Set default session to
.IR sessid .
.LI \f3\(miX\fP
Set default session to the X session of $DISPLAY.
.LI \f3\(miw\0\f2n\f1
Set global timeout to
.I n
seconds.
.LI \f3\(mil\0\f2n\f1
Set
.BR tttrace 
.BR dtterm 
saveLines to
.I n
lines.
.LI \f3\(mio\0\f2snoopfile\f1
.br
Log snoop output to
.IR snoopfile .
.LI \f3\(mio\0\f2tracefile\f1
.br
Log API tracing to
.IR tracefile .
.PP
.RE
.nr )J 0
.SH OPERANDS
The following operands are supported:
.PP
.RS 3
.nr )J 6
.LI \f2command\0[args]\fP
.br
Invoke
.I command
[ with
.I args
] and snoop its ToolTalk API calls.
.PP
.RE
.nr )J 0
.SH RESOURCES
The main widgets that make up the 
.BR ttsnoop 
hierarchy are shown under this
heading to aid in specifying resources.
The widget instance name is shown first,
followed by the widget class name in parentheses.
Indentation indicates hierarchical structure.
.PP
.sp -1
.RS 5
.ta 4m +4m +4m +4m +4m +4m +4m
.nf
.ft 3
Ttsnoop
    ttsnoopWin (topLevelShellWidgetClass)
    dtb_ttsnoop_ttsnoop_pane_pane (DtTerm)
    dtb_message_props_message_props (DialogShell)
    dtb_pattern_props_pattern_props (DialogShell)
    dtb_api_tracer_tracer (DialogShell)
        dtb_api_tracer_trace_pane_obj_pane (DtTerm)
.PP
.ft 1
.fi
.RE
.SH STDIN
Not used.
.SH "INPUT FILES"
None.
.SH "ASYNCHRONOUS EVENTS"
\f3ToolTalk\fP \f3Messages\fP
.PP
If
.I command
is omitted,
.BR ttsnoop 
by default registers a pattern in the default scope to observe all
messages, and prints all observed messages to the terminal pane.
.BR ttsnoop 
can be made to register the handler patterns described in
.BR ttdt_session_join (3),
.BR ttdt_file_join (3),
.BR ttdt_message_accept (3),
.BR ttdt_subcontract_manage (3),
and
.BR ttmedia_ptype_declare (3).
.BR ttsnoop 
can interactively create and register any ToolTalk handler pattern.
The user can install a pattern callback to open the matched messages
in a dialog window, or automatically accept, reject, reply, or fail it.
.SH STDOUT
Not used.
.SH STDERR
Errors encountered during initialization are written to stderr.
After initialization, stderr is not used.
.SH "OUTPUT FILES"
None.
.SH "EXIT STATUS"
The following exit values are returned:
.PP
.RS 3
.nr )J 4
.LI 0
Successful completion.
.LI 1
Could not open X display.
.LI 2
Incorrect command line options.
.LI 3
Could not fork
.IR command .
.LI 4
ToolTalk initialization error.
.LI 5
Caught a fatal signal.
.PP
.RE
.nr )J 0
.SH "NOTES"
.PP
Like any ToolTalk client,
.BR ttsnoop 
can observe multicast messages, but not
.BR \s-1TT_HANDLER\s+1 -\c
addressed messages.  Only
.BR tttrace (1)
can monitor
.BR \s-1TT_HANDLER\s+1 -\c
addressed messages.  Run
.BR tttrace (1)
by choosing "ttsession" from the "Snoop" menu.
.PP
.BR ttsnoop 
allows a message to be opened as long as it has not been destroyed.
However,
.BR ttsnoop 
exposes certain ToolTalk convenience routines --
.BR ttdt_Get_Modified (3),
.BR ttdt_Save (3),
and
.BR ttdt_Revert (3)
-- that send and destroy requests without ever returning them above the API.
When
.BR ttsnoop 
uses these routines to send a request to itself, the message
should not be manipulated after the convenience routines have destroyed it.
.SH FILES
.PP
.RS 3
.nr )J 6
.LI "\f3/usr/dt/app-defaults/$LANG/Ttsnoop\fP"
.br
.BR ttsnoop 
Application Defaults.
.PP
.RE
.nr )J 0
.SH "SEE ALSO"
.na
.BR tttrace (1),
.BR ttsession (1),
.BR tt_type_comp (1),
.BR dttypes (1),
.BR truss (1),
.BR DtTerm (3),
.BR dtactionfile (4)


cde/cde1/ttsnoop/ttsnoop.C
/*** DTB_USER_CODE_START vvv Add file header below vvv ***/
/*** DTB_USER_CODE_END   ^^^ Add file header above ^^^ ***/

/*
 * File: ttsnoop.c
 * Contains: main() and cross-module connections
 *
 * This file was generated by dtcodegen, from project ttsnoop
 *
 * Any text may be added between the DTB_USER_CODE_START and
 * DTB_USER_CODE_END comments (even non-C code). Descriptive comments
 * are provided only as an aid.
 *
 *  ** EDIT ONLY WITHIN SECTIONS MARKED WITH DTB_USER_CODE COMMENTS.  **
 *  ** ALL OTHER MODIFICATIONS WILL BE OVERWRITTEN. DO NOT MODIFY OR  **
 *  ** DELETE THE GENERATED COMMENTS!                                 **
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <X11/Intrinsic.h>
#include <Xm/Xm.h>
#include <Xm/MwmUtil.h>
#include <Xm/Protocols.h>
#include <Dt/Help.h>
#include <Dt/HelpQuickD.h>
#include <Dt/HelpDialog.h>
#include <unistd.h>
#include <Dt/Session.h>
#include "ttsnoop_ui.h"
#include "apiTracer_ui.h"
#include "ttChooser_ui.h"
#include "patternProps_ui.h"
#include "stringChooser_ui.h"
#include "messageProps_ui.h"
#include "sessionChooser_ui.h"
#include "fileChooser_ui.h"
#include "argChooser_ui.h"
#include "callbackChooser_ui.h"
#include "ttsnoop.h"
#include "dtb_utils.h"


/**************************************************************************
 *** DTB_USER_CODE_START
 ***
 *** All necessary header files have been included.
 ***
 *** Add include files, types, macros, externs, and user functions here.
 ***/

#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <time.h>
#include <iostream.h>
#include <strstream.h>
#include <fstream.h>
#include <Dt/Term.h>
#include <Tt/tt_c.h>
#include <Tt/tttk.h>
#include "tt_c++.h"
#include "DtTt.h"

//#include <Xm/ColorObj.h>
//#include <Xm/ColorObjP.h>
//extern "C" { Pixmap XmeGetMask( Screen *, const char * ); }

Boolean		optImmediateTracing	= False;
Boolean		optMapOnOutput		= False;
Boolean		optImmediateSnooping	= True;
Boolean		optImmediateTtOpen	= True;
Boolean		optImmediateXSession	= False;
char *		optImmediateSession	= 0;
Tt_pattern	snoopPat		= 0;
Boolean		snoopPatIsRegistered	= False;
char *		snoopFile		= 0;
Boolean		unlinkSnoopFile		= True;
char *		traceFile		= 0;
Boolean		unlinkTraceFile		= True;
char *		traceScript		= 0;
int		globalTimeout		= 20000;
unsigned int	globalSaveLines		= 5000;
const char *	globalVersionString	= "1.0";
char **		scopeFiles		= 0;
unsigned int	scopeFilesCount		= 0;
char **		ops			= 0;
unsigned int	opsCount		= 0;
char **		senders			= 0;
unsigned int	sendersCount		= 0;
char *		vtype			= 0;
pid_t		snoopedPid		= -1;
char **		snoopedArgs		= 0;
unsigned int	snoopedArgsCount	= 0;
char *		optTraceScript		= 0;
String		apiTracerArgv[ 10 ];
String		snooperArgv[ 10 ];
ostrstream	tttraceCmd;
ofstream	snoopStream;

// Xt squats on -tf ?! XXX
const char Usage[] =
"ttsnoop - ToolTalk graphical user interface\n"
"Usage: ttsnoop [options] [-F scopefile] [-< procid] [-v media] [-m op]\n"
"       ttsnoop [options] -n|-N\n"
"       ttsnoop [options] [-e script] command [args]\n"
"\n"
"-F scopefile	scope initial pattern also to scopefile\n"
"-\\< \"procid\"	limit initial pattern to messages from procid\n"
"-v mediaType	limit initial pattern to messages for mediaType\n"
"-m op		limit initial pattern to given op\n"
"\n"
"-n		skip initial pattern\n"
"-N		skip initial ttdt_open(), also\n"
"\n"
"command [args]	invoke command [with args] and snoop its TT API calls\n"
"-e script	Take script as a tttrace setting.  See tttracefile(4).\n"
"\n"
"options ::=    [-Tu] [-S sessid] [-w n] [-l n] [-o snoopfile] [-O tracefile]\n"
"-T		trace (even initial) TT API calls made by ttsnoop\n"
"-u		map (de-iconify) on snoop output\n"
"-S sessid	set default session to sessid\n"
"-X		set default session to tt_X_session of $DISPLAY\n"
"-w n		set global timeout to n seconds [default: 20]\n"
"-l n		set tttrace dtterm saveLines to n lines [default: 5000]\n"
"-o snoopfile	log snoop output to snoopfile\n"
"-O tracefile	log api tracing to tracefile\n"
"\n"
"-xrm \"*DtTerm.saveLines: n\"\n"
"-xrm \"*DtTerm.rows: n\"\n"
"-xrm \"*DtTerm.columns: n\"\n"
"-xrm \"*DtTerm.mapOnOutput: true\"\n";

static void
cleanUp()
{
	if (unlinkSnoopFile) {
		if (unlink( snoopFile ) == -1) {
			clog << "ttsnoop: unlink( \"" << snoopFile;
			clog << "\"): " << strerror( errno ) << endl;
		} else {
			unlinkSnoopFile = False;
		}
	}
	if (unlinkTraceFile) {
		if (unlink( traceFile ) == -1) {
			clog << "ttsnoop: unlink( \"" << traceFile;
			clog << "\"): " << strerror( errno ) << endl;
		} else {
			unlinkTraceFile = False;
		}
	}
}

static void
signalHandler(
	int sig
)
{
	pid_t child;
	int status;
	switch (sig) {
	    case SIGHUP:
	    case SIGINT:
	    case SIGQUIT:
	    case SIGPIPE:
	    case SIGTERM:
		exit( 5 );
	    case SIGCHLD:
		child = waitpid( snoopedPid, &status, WNOHANG );
		if ((child > 0) && WIFEXITED( status )) {
			snoopStream << endl << endl << "SIGCHLD: WEXITSTATUS=="
				<< WEXITSTATUS(status) << ": "
				<< tttraceCmd.str() << endl << endl;
		}
		break;
	    case SIGCONT:
		break;
	}
}

#if defined(SVR4) || defined(aix) || defined(hpux)
#if !defined(SIG_PF)
typedef void (*sig_pf_t)(int);
#define SIG_PF sig_pf_t
#endif
#endif

int
_tt_sigset(
	int     sig,
	SIG_PF  handler )
{
#if defined(hpux)
	struct sigaction act;
	act.sa_handler = handler;
	sigemptyset(&act.sa_mask);
	act.sa_flags = 0;
	return 0==sigaction(sig, &act, NULL);
#elif defined(OPT_BSD_SIGNAL)
	return SIG_ERR!=signal(sig, handler);
#else
	return SIG_ERR!=sigset(sig, handler);
#endif
}

static void
installSignalHandler()
{
    _tt_sigset( SIGHUP, signalHandler );
    _tt_sigset( SIGINT, signalHandler );
    _tt_sigset( SIGQUIT, signalHandler );
    _tt_sigset( SIGPIPE, signalHandler );
    _tt_sigset( SIGTERM, signalHandler );
    _tt_sigset( SIGCHLD, signalHandler );
    _tt_sigset( SIGCONT, signalHandler );
    atexit( cleanUp );
}

void
snoopIt(
	const char *callBackType,
	void *callBack,
	Tt_message msg,
	Tt_pattern pat,
	Boolean printPat
)
{
	char timeBuf[ 16 ];
	time_t now = time( 0 );
	strftime( timeBuf, sizeof( timeBuf ), "%T", localtime( &now ));
	snoopStream << timeBuf << " (" << callBackType << ")" << callBack;
	snoopStream << "( " << "(Tt_message)" << (void *)msg << ", ";
	if (printPat) {
		snoopStream << "(Tt_pattern)" << (void *)pat;
	} else {
		snoopStream << "...";
	}
	snoopStream << " ): ";
	char *msgString = tt_message_print( msg );
	Tt_status status = tt_ptr_error( msgString );
	if (status == TT_OK) {
		snoopStream << endl << msgString << endl;
	} else {
		snoopStream << status << endl;
	}
	tt_free( msgString );
}

static Tt_callback_action
justSnoopIt(
	Tt_message msg,
	Tt_pattern pat
)
{
	DtTtCreated( DTTT_MESSAGE, msg ); // XXX bad idea?
	snoopIt( "Tt_callback_action", justSnoopIt, msg, pat, True );
	return TT_CALLBACK_PROCESSED;
}

Tt_message
_DtTtMediaLoadPatCb(
	Tt_message	msg,
	void           *clientdata,
	Tttk_op		op,
	Tt_status	diagnosis,
	unsigned char  *contents,
	int             len,
	char           *file,
	char           *docname
)
{
	tt_free( (caddr_t)contents );
	tt_free( file );
	tt_free( docname );
	snoopIt( "Ttmedia_load_pat_cb", _DtTtMediaLoadPatCb, msg );
	return 0;
}

/*** DTB_USER_CODE_END
 ***
 *** End of user code section
 ***
 **************************************************************************/



/*
 * command line options...
 */
static XrmOptionDescRec optionDescList[] = {
    {"-session", "*session", XrmoptionSepArg, (XPointer)NULL}
    
    /*** DTB_USER_CODE_START vvv Add structure fields below vvv ***/
    /*** DTB_USER_CODE_END   ^^^ Add structure fields above ^^^ ***/
};

/*
 * Application Resources
 */
static XtResource resources[] = {
    {"session", "Session", XtRString, sizeof(String),
        XtOffsetOf(DtbAppResourceRec, session_file), XtRImmediate, (XtPointer)NULL}
    
    /*** DTB_USER_CODE_START vvv Add structure fields below vvv ***/
    /*** DTB_USER_CODE_END   ^^^ Add structure fields above ^^^ ***/
};

DtbAppResourceRec	dtb_app_resource_rec;


/*
 * main for application ttsnoop
 */
int
main(int argc, char **argv)
{
    Widget		toplevel = NULL;
    Display		*display = NULL;
    XtAppContext	app;
    Atom		save_yourself_atom;

    Pixmap	icon_pixmap = NULL;
    Pixmap	icon_mask_pixmap = NULL;
    
    /**************************************************************************
     *** DTB_USER_CODE_START
     ***
     ***  No initialization has been done.
     ***
     *** Add local variables and code.
     ***/

    DtTermInitialize();

    /*** DTB_USER_CODE_END
     ***
     *** End of user code section
     ***
     **************************************************************************/
    

    toplevel = XtVaAppInitialize(&app, "Ttsnoop",
        optionDescList, XtNumber(optionDescList),
        &argc, argv, (String *)NULL,
        (ArgList) NULL, 0);

    /*
     * Get display and verify initialization was successful.
     */
    if (toplevel != NULL)
    {
        display = XtDisplayOfObject(toplevel);
    }
    if (display == NULL)
    {
        fprintf(stderr, "Could not open display.");
        exit(1);
    }
    
    /*
     * Save the toplevel widget so it can be fetched later as needed.
     */
    dtb_save_toplevel_widget(toplevel);
    
    /*
     * Save the command used to invoke the application.
     */
    dtb_save_command(argv[0]);
    
    XtGetApplicationResources(toplevel, (XtPointer)&dtb_app_resource_rec,
    	resources, XtNumber(resources), (Arg *)NULL, 0);
    
    
    /**************************************************************************
     *** DTB_USER_CODE_START
     ***
     ***      A connection to the X server has been established, and all
     *** initialization has been done.
     ***
     ***  Add extra initialization code after this comment.
     ***/

    /*
    PixelSet	pixels[MAX_NUM_COLORS];
    int		colorUse;
    short	act, inact, prim, second;
    _XmGetPixelData( DefaultScreen( display ), &colorUse, pixels,
		     &act, &inact, &prim, &second );
    Pixmap pixmap = XmGetPixmap( DefaultScreenOfDisplay( display ),
				 "DtTtsnp.l", pixels[1].fg, pixels[1].bg );
    XtVaSetValues( toplevel, XmNiconPixmap, pixmap, NULL) ;
    pixmap = XmeGetMask( XtScreen( toplevel ), "DtTtsnp.l" );
    XtVaSetValues( toplevel, XmNiconMask, pixmap, NULL) ;
    */
    Boolean committed2Snooping = False;
    int c;
    while ((c = getopt(argc, argv, "?F:<:v:nNe:TuS:Xw:l:o:O:")) != -1) {
	    switch (c) {
		case 'F':
		    if (! optImmediateSnooping) {
			    clog << Usage; exit( 2 );
		    }
		    committed2Snooping = 1;
		    listAppend( scopeFiles, scopeFilesCount, char *, optarg );
		    break;
		case '<':
		    if (! optImmediateSnooping) {
			    clog << Usage; exit( 2 );
		    }
		    committed2Snooping = 1;
		    listAppend( senders, sendersCount, char *, optarg );
		    break;
		case 'v':
		    if ((! optImmediateSnooping) || (vtype != 0)) {
			    clog << Usage; exit( 2 );
		    }
		    committed2Snooping = 1;
		    vtype = optarg;
		    break;
		case 'm':
		    if (! optImmediateSnooping) {
			    clog << Usage; exit( 2 );
		    }
		    committed2Snooping = 1;
		    listAppend( ops, opsCount, char *, optarg );
		    break;
		case 'n':
		    if (committed2Snooping) {
			    clog << Usage; exit( 2 );
		    }
		    optImmediateSnooping = False;
		    break;
		case 'N':
		    if (committed2Snooping) {
			    clog << Usage; exit( 2 );
		    }
		    optImmediateSnooping = False;
		    optImmediateTtOpen = False;
		    break;
		case 'e':
		    optTraceScript = optarg;
		    break;
		case 'T':
		    optImmediateTracing = True;
		    break;
		case 'u':
		    optMapOnOutput = True;
		    break;
		case 'S':
		    if (optImmediateSession != 0) {
			    clog << Usage; exit( 2 );
		    }
		    optImmediateSession = optarg;
		    break;
		case 'X':
		    if (optImmediateSession != 0) {
			    clog << Usage; exit( 2 );
		    }
		    optImmediateXSession = True;
		    break;
		case 'w':
		    globalTimeout = atoi( optarg );
		    if (globalTimeout > 0) {
			    globalTimeout *= 1000;
		    }
		    break;
		case 'l':
		    globalSaveLines = atoi( optarg );
		    break;
		case 'o':
		    if (snoopFile != 0) {
			    clog << Usage; exit( 2 );
		    }
		    snoopFile = optarg;
		    unlinkSnoopFile = False;
		    break;
		case 'O':
		    if (traceFile != 0) {
			    clog << Usage; exit( 2 );
		    }
		    traceFile = optarg;
		    unlinkTraceFile = False;
		    break;
		case '?':
		default:
		    clog << Usage; exit( 2 );
	    }
    }

    installSignalHandler();
    if (traceFile == 0) {
	    //
	    // Set up fifo for trace output
	    //
	    traceFile = tempnam( 0, "ttsnt" );
	    if (mkfifo( traceFile, S_IWUSR | S_IRUSR ) == -1) {
		    clog << "ttsnoop: mkfifo( \"" << traceFile << "\" ) = ";
		    clog << strerror( errno ) << endl;
	    }
    }
    apiTracerArgv[ 0 ] = "tail";
    apiTracerArgv[ 1 ] = "+0f";
    apiTracerArgv[ 2 ] = traceFile;
    if (snoopFile == 0) {
	    //
	    // Set up fifo for snoop output
	    //
	    snoopFile = tempnam( 0, "ttsnp" );
	    if (mkfifo( snoopFile, S_IWUSR | S_IRUSR ) == -1) {
		    clog << "ttsnoop: mkfifo( \"" << snoopFile << "\" ) = ";
		    clog << strerror( errno ) << endl;
	    }
    }
    snooperArgv[ 0 ] = "tail";
    snooperArgv[ 1 ] = "+0f";
    snooperArgv[ 2 ] = snoopFile;

    if (optind < argc) {
	    if (committed2Snooping) {
		    clog << Usage; exit( 2 );
	    }
	    optImmediateSnooping = False;
	    optImmediateTtOpen = False;
	    listAppend( snoopedArgs, snoopedArgsCount, char *, "tttrace" );
	    listAppend( snoopedArgs, snoopedArgsCount, char *, "-o" );
	    listAppend( snoopedArgs, snoopedArgsCount, char *, snoopFile );
	    tttraceCmd << "tttrace -o " << snoopFile;
	    if (optTraceScript != 0) {
		    listAppend( snoopedArgs, snoopedArgsCount, char *, "-e" );
		    listAppend( snoopedArgs, snoopedArgsCount, char *,
			        optTraceScript );
		    tttraceCmd << " -e \"" << optTraceScript << "\"";
	    }
	    for (int i = optind; i < argc; i++) {
		    listAppend( snoopedArgs, snoopedArgsCount, char *,
			        argv[i] );
		    tttraceCmd << " " << argv[i];
	    }
	    tttraceCmd << ends;
	    listAppend( snoopedArgs, snoopedArgsCount, char *, 0 );
    }

#if defined(aix)
#define AIX_STRING_LIST		(const char **)
#else
#define AIX_STRING_LIST
#endif

    if (snoopedArgsCount > 0) {
	    snoopedPid = fork();
	    switch (snoopedPid) {
		case -1:
		    clog << "ttsnoop: fork(): " << strerror( errno ) << endl;
		    exit( 3 );
		case 0:
		    execvp( snoopedArgs[0], AIX_STRING_LIST snoopedArgs );
		    clog << "ttsnoop: execvp(): " << strerror( errno ) << endl;
		    exit( 3 );
	    }
	    installSignalHandler();
    }
    
    /*** DTB_USER_CODE_END
     ***
     *** End of user code section
     ***
     **************************************************************************/
    

    
    /*
     * Initialize all global variables.
     */
    dtbTtsnoopTtsnoopWinInfo_clear(&dtb_ttsnoop_ttsnoop_win);
    dtbApiTracerTracerInfo_clear(&dtb_api_tracer_tracer);
    dtbTtChooserChooserInfo_clear(&dtb_tt_chooser_chooser);
    dtbPatternPropsPatternPropsInfo_clear(&dtb_pattern_props_pattern_props);
    dtbStringChooserStringChooserInfo_clear(&dtb_string_chooser_string_chooser);
    dtbMessagePropsMessagePropsInfo_clear(&dtb_message_props_message_props);
    dtbSessionChooserSessionChooserInfo_clear(&dtb_session_chooser_session_chooser);
    dtbFileChooserFchooserInfo_clear(&dtb_file_chooser_fchooser);
    dtbArgChooserArgChooserInfo_clear(&dtb_arg_chooser_arg_chooser);
    dtbCallbackChooserCallbackChooserInfo_clear(&dtb_callback_chooser_callback_chooser);
    
    /*
     * Set up the application's root window.
     */
    dtb_ttsnoop_ttsnoop_win.ttsnoopWin = toplevel;
    dtb_cvt_image_file_to_pixmap(dtb_ttsnoop_ttsnoop_win.ttsnoopWin,
        "DtTtsnp.l", &icon_pixmap);
    dtb_cvt_image_file_to_pixmap(dtb_ttsnoop_ttsnoop_win.ttsnoopWin,
        "DtTtsnp.l_m", &icon_mask_pixmap);
    XtVaSetValues(dtb_ttsnoop_ttsnoop_win.ttsnoopWin,
        XmNallowShellResize, True,
        XmNtitle, "ttsnoop",
        XmNiconMask, icon_mask_pixmap,
        XmNiconPixmap, icon_pixmap,
        XmNbackground, dtb_cvt_string_to_pixel(dtb_ttsnoop_ttsnoop_win.ttsnoopWin, "white"),
        NULL);
    
    dtb_ttsnoop_ttsnoop_win_initialize(&(dtb_ttsnoop_ttsnoop_win), dtb_get_toplevel_widget());
    
    /*
     * Map any initially-visible windows
     */
    
    save_yourself_atom = XmInternAtom(XtDisplay(toplevel),
    	"WM_SAVE_YOURSELF", False);
    
    dtb_set_client_session_saveCB((DtbClientSessionSaveCB)NULL);
    
    XmAddWMProtocolCallback(toplevel, save_yourself_atom,
    	dtb_session_save, (XtPointer)NULL);
    
    
    /**************************************************************************
     *** DTB_USER_CODE_START
     ***
     ***      All initially-mapped widgets have been created, but not
     *** realized. Set resources on widgets, or perform other operations
     *** that must be completed before the toplevel widget is
     *** realized.
     ***/
    
    /*** DTB_USER_CODE_END
     ***
     *** End of user code section
     ***
     **************************************************************************/
    

    XtRealizeWidget(toplevel);

    
    /**************************************************************************
     *** DTB_USER_CODE_START
     ***
     ***      The initially-mapped widgets have all been realized, and
     *** the Xt main loop is about to be entered.
     ***/

    installSignalHandler();
    if (snoopedArgsCount > 0) {
	    DtTtSetLabel( dtb_ttsnoop_ttsnoop_win.ttsnoopWin_label,
			  tttraceCmd.str() );
    }
    Tt_status status;
    snoopStream.open( snoopFile,
		      getenv( "libC_1187592_fixed" ) ? ios::app : ios::out );
    if (! snoopStream.good()) {
	    clog << "ttsnoop: " << snoopFile << ": "
		 << strerror( errno ) << endl;
    }
    ostrstream envStr;
    envStr << "TT_TRACE_SCRIPT=> ";
    envStr << traceFile << ends;
    traceScript = envStr.str();
    if (optImmediateTracing) {
	    turnOnTracing( 0, 0, 0 );
    }
    if (optImmediateXSession) {
	    char *display = getenv( "DISPLAY" );
	    optImmediateSession = tt_X_session( display );
	    status = tt_ptr_error( optImmediateSession );
	    if (tt_is_err( status )) {
		    clog << "ttsnoop: tt_X_session( ";
		    if (display == 0) {
			    clog << "0";
		    } else {
			    clog << "\"" << display << "\"";
		    }
		    clog << " ) = " << (void *)optImmediateSession;
		    clog << " (" << status << ")" << endl;
		    exit( 4 );
	    }
    }
    if (optImmediateSession != 0) {
	    status = tt_default_session_set( optImmediateSession );
	    if (tt_is_err( status )) {
		    clog << "ttsnoop: tt_default_session_set( \"";
		    clog << optImmediateSession << "\" ) = ";
		    clog << status << endl;
		    exit( 4 );
	    }
    }
    if (optImmediateTtOpen) {
	    int fd;
	    char *procid = ttdt_open( &fd, "Ttsnoop", "CDE",
				      globalVersionString, 1 );
	    DtTtSetLabel( dtb_ttsnoop_ttsnoop_win.ttsnoopWin_label,
			  "ttdt_open()", procid );
	    status = tt_ptr_error( procid );
	    if (tt_is_err( status )) {
		    clog << "ttsnoop: ttdt_open() = " << status << endl;
		    exit( 4 );
	    }
	    XtInputId id = XtAppAddInput( app, fd, (XtPointer)XtInputReadMask,
					  tttk_Xt_input_handler, procid );
	    DtTtCreated( DTTT_PROCID, procid, (void *)id );
	    if (optImmediateSnooping) {
		    Tt_pattern pat = tt_pattern_create();
		    tt_pattern_category_set( pat, TT_OBSERVE );
		    if (scopeFilesCount > 0) {
			    tt_pattern_scope_add( pat, TT_BOTH );
			    for (int i = 0; i < scopeFilesCount; i++) {
				    tt_pattern_file_add( pat, scopeFiles[i] );
			    }
		    } else {
			    tt_pattern_scope_add( pat, TT_SESSION );
		    }
		    char *sess = tt_default_session();
		    tt_pattern_session_add( pat, sess );
		    tt_free( sess );
		    for (int i = 0; i < opsCount; i++) {
			    tt_pattern_op_add( pat, ops[i] );
		    }
		    for (i = 0; i < sendersCount; i++) {
			    tt_pattern_sender_add( pat, senders[i] );
		    }
		    if (vtype != 0) {
			    tt_pattern_arg_add( pat, TT_MODE_UNDEFINED,
					        vtype, 0 );
		    }
		    tt_pattern_callback_add( pat, justSnoopIt );
		    status = tt_pattern_register( pat );
		    DtTtSetLabel( dtb_ttsnoop_ttsnoop_win.ttsnoopWin_label,
				  "tt_pattern_register()", pat );
		    if (tt_is_err( status )) {
			    clog << "ttsnoop: tt_pattern_register() = ";
			    clog << status << endl;
			    exit( 4 );
		    }
		    DtTtCreated( DTTT_PATTERN, pat );
		    snoopPat = pat;
		    snoopPatIsRegistered = True;
	    }
    }
    if (snoopPat == 0) {
	    XtSetSensitive( dtb_ttsnoop_ttsnoop_win.
		menubar_Snoop_item_Snoop_menu_items.Off_item, False );
    }
    DtTtSetLabel( dtb_ttsnoop_ttsnoop_win.
		  menubar_Snoop_item_Snoop_menu_items.Off_item,
		  snoopPatIsRegistered ? "Off" : "On" );
    if (optMapOnOutput) {
	    XtVaSetValues( dtb_ttsnoop_ttsnoop_win.ttsnoopPane,
			   DtNmapOnOutput, optMapOnOutput, 0 );
    }
    installSignalHandler();
    
    /*** DTB_USER_CODE_END
     ***
     *** End of user code section
     ***
     **************************************************************************/
    

    
    /*
     * Enter event loop
     */
    XtAppMainLoop(app);
    exit(0);
}


/**************************************************************************
 *** DTB_USER_CODE_START
 ***
 *** All automatically-generated data and functions have been defined.
 ***
 *** Add new functions here, or at the top of the file.
 ***/

/*** DTB_USER_CODE_END
 ***
 *** End of user code section
 ***
 **************************************************************************/


